fix(provider): support local file paths for custom providers

sredfern opened the pull request:
# PR: Fix local file path support for custom providers

## Summary

Fixes a bug where OpenCode cannot load custom providers from local file paths. Previously, all provider paths were treated as npm packages, causing `BunInstallFailedError` when trying to load local providers.

This implementation uses the same pattern as plugin loading for consistency.

## Problem

When configuring a custom provider with a local path:

```json
{
  "provider": {
    "my-provider": {
      "npm": "file://./my-provider"
    }
  }
}
```

OpenCode would attempt to install it via npm:
```bash
bun add --force --exact file://./my-provider@latest
```

This fails with:
```
BunInstallFailedError: package 'file://./my-provider' not found in registry
```

## Root Cause

In `packages/opencode/src/provider/provider.ts`, the `getSDK()` function always called `BunProc.install(pkg, "latest")` without checking if `pkg` was a local path or npm package name.

```typescript
// Before (broken):
const installedPath = await BunProc.install(pkg, "latest")  // ❌ Always tries npm install
```

## Solution

Add detection for `file://` prefix before attempting npm installation, matching the existing pattern in `src/plugin/index.ts`:

```typescript
// After (fixed) - matches plugin loading pattern:
let installedPath: string
if (!pkg.startsWith("file://")) {
  installedPath = await BunProc.install(pkg, "latest")
} else {
  log.info("loading local provider", { pkg })
  installedPath = pkg
}
```

The fix:
1. **Checks for `file://` prefix** - same as plugin loading (consistent with existing codebase)
2. **If no prefix** → npm install (existing behavior)
3. **If has prefix** → use path directly for import

## Testing

### Before Fix (Bug Confirmed)
```bash
# With opencode v1.0.55 (before fix)
$ opencode run -m "my-provider/model" "test"

ProviderInitError: ProviderInitError
BunInstallFailedError: package 'file://./my-provider' not found in registry
```

### After Fix (Working)
```bash
# With fix applied
$ opencode run -m "my-provider/model" "test"
✓ Works correctly - provider loaded from local path
```

## Use Cases Enabled

1. **Development workflow**: Test custom providers locally without publishing
2. **Private/internal providers**: Use company-internal providers without npm registry
3. **Monorepo packages**: Reference sibling packages in monorepo
4. **Custom integrations**: Develop provider for proprietary LLM gateway

## Configuration Example

```json
{
  "provider": {
    "my-custom-provider": {
      "name": "My Custom Provider",
      "npm": "file://./providers/my-provider",
      "models": {
        "custom-model": { "name": "Custom Model" }
      }
    },
    "npm-package": {
      "npm": "@ai-sdk/anthropic"  // Still works as before
    }
  }
}
```

**Note**: Must use `file://` prefix for local paths (matches plugin pattern)

## Implementation Details

### Matches Existing Plugin Pattern

This uses the **exact same approach** as plugin loading in `src/plugin/index.ts:36`:

```typescript
// Plugin loading (existing):
if (!plugin.startsWith("file://")) {
  plugin = await BunProc.install(pkg, version)
}

// Provider loading (this PR):
if (!pkg.startsWith("file://")) {
  installedPath = await BunProc.install(pkg, "latest")
}
```

**Benefits:**
- ✅ Consistent with existing codebase patterns
- ✅ Simple and easy to understand
- ✅ Minimal code changes (6 lines)

## Backwards Compatibility

✅ **Fully backwards compatible**

- Existing npm package configurations work unchanged
- No breaking changes to API or configuration format
- Falls back to original behavior for non-`file://` paths

## Related

- Addresses issue where users cannot develop custom providers locally
- Uses same pattern as plugin loading for consistency
- Requested by @rekram1-node to match existing codebase patterns

## Files Changed

- `packages/opencode/src/provider/provider.ts` - Added `file://` prefix detection in `getSDK()` function

## Commits

- `3839d6a5` - Initial fix with multiple path formats
- `2766b222` - Updated to match plugin loading pattern (current)

## How to Test

1. Create a simple test provider:

```bash
mkdir -p ./test-provider
cat > ./test-provider/index.js << 'EOF'
export function createTestProvider() {
  return {
    languageModel() {
      return {
        provider: 'test',
        modelId: 'test-model'
      }
    }
  }
}
EOF
```

2. Configure OpenCode:

```json
{
  "provider": {
    "test-local": {
      "name": "Test Local Provider",
      "npm": "file://./test-provider",
      "models": {
        "test-model": { "name": "Test Model" }
      }
    }
  }
}
```

3. Test loading:

```bash
opencode models | grep test-local
# Should show: test-local/test-model

opencode run -m "test-local/test-model" "test"
# Should work without BunInstallFailedError
```

## Checklist

- [x] Bug confirmed on latest `dev` branch
- [x] Fix implemented matching existing plugin pattern
- [x] No breaking changes
- [x] Backwards compatible with existing configs
- [x] Matches codebase style and patterns
- [x] Tested locally
- [x] Ready for review

---

**Reviewer Notes:**
- Minimal change: only 6 lines (4 insertions, 1 deletion after refactor)
- Uses exact same pattern as plugin loading for consistency
- Simple logic: if not `file://`, npm install; otherwise use path
- No changes to error handling, caching, or other provider loading logic

sredfern committed:
fix(provider): support local file paths for custom providers

Add detection for local file paths (absolute paths starting with '/',
relative paths starting with './', and file:// URLs) to skip npm
installation and directly import the module. This enables custom
providers to be loaded from local files without requiring npm publication.

Previously, BunProc.install() was always called regardless of whether
the path was a local file or npm package. This caused errors when
trying to fetch local paths from npm registry, resulting in 404 errors
and BunInstallFailedError.

This fix:
- Detects local paths using startsWith() checks for '/', './', 'file:'
- Resolves relative paths relative to Instance.directory (project root)
- Strips 'file:' protocol prefix and resolves path correctly
- Falls back to BunProc.install() for npm packages (existing behavior)

Enables use cases:
- Development of custom providers locally
- Private/internal providers without publishing to npm
- Testing provider implementations before publication
- Using providers from monorepo sibling packages

Example configurations that now work:
- "npm": "/absolute/path/to/provider"
- "npm": "./relative/path/to/provider"
- "npm": "file://./providers/custom-provider"

Fixes issue where local provider paths would fail with:
  BunInstallFailedError: package '/path/to/provider' not found in registry

sredfern commented:
Just some context - The company I work for runs a custom llm gateway. This fix would mean we can use OpenCode internally as an agent for different custom ai harness applications.

rekram1-node commented:
@sredfern can we just use this same condition we have elsewhere and check for the file prefix?

example: https://github.com/sst/opencode/blob/b9394703026bffb984297634b7d16db3c4555def/packages/opencode/src/plugin/index.ts#L36

sredfern committed:
Use same file:// pattern as plugin loading

Updated to match the existing pattern in src/plugin/index.ts where we
only check for file:// prefix. This is simpler and more consistent
with the rest of the codebase.

- Check if pkg starts with 'file://'
- If not, treat as npm package and install
- If yes, use path directly for import

This matches how plugins are loaded and keeps the codebase consistent.

sredfern commented:
Hey @rekram1-node - This is done. Please review.

sredfern committed:
Merge branch 'dev' into fix/local-provider-support

rekram1-node commented:
merged thanks!
